### DISCUSSION TITLE
Is there a way to get a builder for a struct ?

### DISCUSSION THREAD

**Discussion by cameronbraid** [idanarye/rust-typed-builder#99] [Q&A] [answered]
For example :

```
#[derive(Clone, Debug, TypedBuilder)]
pub struct Foo {
  width: usize,
  height: usize,
}

let foo = Foo::builder().width(0).height(0).build();
let new_foo = foo.to_builder().height(1).build(); // <--- the to_builder method in here
// new_foo.width is 0
```

**Comment by idanarye** [comment, marked as answer]
This crate does not support it, and I don't think it should.

Typed Builder's first and foremost purpose is to allow fields with default values to get their default values while ensuring, at compile time, that all the mandatory fields are set. If you create a builder from a struct, all the fields are already set - making this enforcement redundant.

Typed Builder also verifies that no field is set in the builder more than once. This is actually a problem in your case - if the builder is initialized by the original struct, then any field you set in it will be set a second time. Of course, this can be resolved by deciding that fields can be set once in the builder in addition to being initialized from the struct, but because of the architecture of Typed Builder where being able to store a value is the very indicator of whether or not it was set, this is going to be very tricky to implement.

Also, one of the reasons for this crate was a limitation of the `..Default::default()` pattern - that it forces you to define a default value for all fields. If you already have a a struct, Rust has a builtin way to do what you want:

```rust
let new_foo = Foo {
    height: 1,
    ..foo
};
```

**Comment by cameronbraid** [comment]
Thanks. I appreciate your detailed explanation and totally agree with you. 

And thanks for this great library
